File: cliphist_picker/dedup.py
from .models import ClipboardEntry, EntryType, RawEntry


def deduplicate(
    raw_entries: list[RawEntry],
    id_threshold: int = 3,
    max_entries: int = 50,
) -> list[ClipboardEntry]:
    """Group adjacent text+image entries from the same copy operation.

    cliphist list returns entries most-recent-first (descending ID).
    When a single copy stores both text and image, their IDs are
    within ``id_threshold`` and they appear adjacent.

    Walk linearly: at index *i*, peek at the next unconsumed index *j*.
    If one is binary and the other text with IDs close enough, merge
    into a MERGED entry. Otherwise emit standalone.
    """
    result: list[ClipboardEntry] = []
    consumed: set[int] = set()
    i = 0

    while i < len(raw_entries) and len(result) < max_entries:
        if i in consumed:
            i += 1
            continue

        entry = raw_entries[i]

        # Find the next unconsumed entry
        j = i + 1
        while j < len(raw_entries) and j in consumed:
            j += 1

        merged = False
        if j < len(raw_entries):
            neighbor = raw_entries[j]
            id_gap = abs(entry.id - neighbor.id)

            if id_gap <= id_threshold and entry.is_binary != neighbor.is_binary:
                if entry.is_binary:
                    img_entry, txt_entry = entry, neighbor
                else:
                    txt_entry, img_entry = entry, neighbor

                result.append(ClipboardEntry(
                    entry_type=EntryType.MERGED,
                    text_id=txt_entry.id,
                    text_preview=txt_entry.preview,
                    image_id=img_entry.id,
                    image_size=img_entry.binary_size,
                    image_format=img_entry.binary_format,
                    image_dims=img_entry.binary_dims,
                    primary_raw_line=txt_entry.raw_line,
                    image_raw_line=img_entry.raw_line,
                ))
                consumed.add(i)
                consumed.add(j)
                merged = True

        if not merged:
            if entry.is_binary:
                result.append(ClipboardEntry(
                    entry_type=EntryType.IMAGE,
                    image_id=entry.id,
                    image_size=entry.binary_size,
                    image_format=entry.binary_format,
                    image_dims=entry.binary_dims,
                    primary_raw_line=entry.raw_line,
                    image_raw_line=entry.raw_line,
                ))
            else:
                result.append(ClipboardEntry(
                    entry_type=EntryType.TEXT,
                    text_id=entry.id,
                    text_preview=entry.preview,
                    primary_raw_line=entry.raw_line,
                ))
            consumed.add(i)

        i += 1

    return result

File: src/dedup.c
#include "dedup.h"
#include "clipboard.h"

#include <stdlib.h>
#include <string.h>

ClipboardEntry *deduplicate(const RawEntry *raw, int raw_count,
                            int max_entries,
                            bool current_clipboard_multi_type,
                            int *out_count) {

    int limit = raw_count < max_entries ? raw_count : max_entries;
    ClipboardEntry *result = g_new0(ClipboardEntry, limit);
    int count = 0;

    /*
     * If the current clipboard holds both text and image MIME types,
     * the most recent copy was a multi-type event (e.g. right-click
     * copy from a browser).  Find the text entry paired with the
     * image at the top and permanently delete it from cliphist.
     */
    int skip_id = -1;
    if (current_clipboard_multi_type && raw_count >= 2) {
        const RawEntry *a = &raw[0];
        const RawEntry *b = &raw[1];
        if (a->is_binary != b->is_binary) {
            const RawEntry *txt = a->is_binary ? b : a;
            skip_id = txt->id;
            clipboard_delete(txt->raw_line);
        }
    }

    for (int i = 0; i < raw_count && count < max_entries; i++) {
        const RawEntry *entry = &raw[i];

        if (entry->id == skip_id)
            continue;

        ClipboardEntry *ce = &result[count];
        if (entry->is_binary) {
            ce->entry_type      = ENTRY_TYPE_IMAGE;
            ce->text_id         = -1;
            ce->image_id        = entry->id;
            ce->image_size      = g_strdup(entry->binary_size);
            ce->image_format    = g_strdup(entry->binary_format);
            ce->image_dims      = g_strdup(entry->binary_dims);
            ce->primary_raw_line = g_strdup(entry->raw_line);
            ce->image_raw_line  = g_strdup(entry->raw_line);
        } else {
            ce->entry_type      = ENTRY_TYPE_TEXT;
            ce->text_id         = entry->id;
            ce->text_preview    = g_strdup(entry->preview);
            ce->image_id        = -1;
            ce->primary_raw_line = g_strdup(entry->raw_line);
        }
        count++;
    }

    *out_count = count;
    return result;
}
